From 3f4ba7c6912c05fb54ec8fb1e3bbd417ff349a97 Mon Sep 17 00:00:00 2001
From: Gareth Davidson <gaz@bitplane.net>
Date: Sun, 6 Sep 2026 16:57:05 +0100
Subject: [PATCH 5/7] Implement TempleOS exception interrupt and variadic ABIs

Select the matching try/untry interface, preserve interrupt function flags and native segments, and emit interrupt returns. Model TempleOS varargs as a stack array and preserve caller cleanup for variadic functions.
---
 Src/AIWNIOS_CodeGen.HC | 50 ++++++++++++++++++++++++++++++++++++------
 Src/AIWNIOS_PrsExp.HC  |  3 +--
 Src/CInit.HC           |  2 +-
 Src/PrsStmt.HC         | 46 ++++++++++++++++++++++++++++++++------
 Src/PrsVar.HC          | 19 +++++++++++-----
 c/aiwn_lexparser.h     |  8 +++++++
 c/arm_backend.c        |  4 ++++
 c/main.c               | 24 ++++++++++++++++++++
 c/optpass.c            |  5 ++++-
 c/parser.c             | 26 ++++++++++++++++++++++
 c/riscv_backend.c      |  4 ++++
 c/x86_64_backend.c     | 39 ++++++++++++++++++++++++++++----
 12 files changed, 203 insertions(+), 27 deletions(-)

diff --git a/Src/AIWNIOS_CodeGen.HC b/Src/AIWNIOS_CodeGen.HC
index 57f4d4c..bd79f3c 100644
--- a/Src/AIWNIOS_CodeGen.HC
+++ b/Src/AIWNIOS_CodeGen.HC
@@ -5,6 +5,9 @@
 //
 #ifdef IMPORT_AIWNIOS_SYMS
 import U0 __HC_CmpCtrl_SetRawAOTAsm(U8 *);
+import U0 __HC_CmpCtrl_UseNativeSegments(U8 *);
+import U0 __HC_CmpCtrl_SetVargs(U8 *);
+import U0 __HC_CmpCtrl_SetInterrupt(U8 *,Bool);
 import U0 __HC_SetAOTRelocBeforeRIP(U8 *,I64); //See aiwn.h(TAKES number of bytes before current RIP the symbol is located at)
 import U8 *__HC_CmpCtrlDel(U8 *);
 import U8 *__HC_ICAdd_Sqr(U8 *);
@@ -87,6 +90,7 @@ import U8 *__HC_Compile(U8*,I64 *sz=NULL,U8 **dbg_info=NULL,CHeapCtrl *);
 import U8 *__HC_CodeMiscStrNew(U8 *,U8*,I64);
 import U8 *__HC_CodeMiscJmpTableNew(U8*,U8**,I64,I64);
 import U8 *__HC_ICAdd_Label(U8*,U8*);
+import U8 *__HC_ICAdd_LabelAddr(U8*,U8*);
 import U8 *__HC_ICAdd_Goto(U8*,U8*);
 import U8 *__HC_ICAdd_GotoIf(U8*,U8*)
 import U8 *__HC_ICAdd_Str(U8*,U8*); //Takes a AIWNIOS code misc
@@ -174,6 +178,8 @@ extern U8 * __HC_ICAdd_AndEq(U8*);
 extern U8 * __HC_ICAdd_OrEq(U8*);
 extern U8 * __HC_ICAdd_XorEq(U8*);
 extern U8 * __HC_ICAdd_ModEq(U8*);
+extern U0 __HC_CmpCtrl_SetInterrupt(U8 *,Bool);
+extern U0 __HC_CmpCtrl_SetVargs(U8 *);
 extern U8 * __HC_ICAdd_FReg(U8*,I64);
 extern U8 * __HC_ICAdd_Ds(U8 *,I64);
 extern U8 * __HC_ICAdd_IReg(U8*,I64 r,I64 rt,I64 ptrs);
@@ -189,6 +195,7 @@ extern U8 *__HC_CodeMiscLabelNew(U8 *,U8 **patch_addr=NULL);
 extern U8 *__HC_CodeMiscStrNew(U8 *,U8*,I64);
 extern U8 *__HC_CodeMiscJmpTableNew(U8*,U8**,I64,I64);
 extern U8 *__HC_ICAdd_Label(U8*,U8*);
+extern U8 *__HC_ICAdd_LabelAddr(U8*,U8*);
 extern U8 *__HC_ICAdd_Goto(U8*,U8*);
 extern U8 *__HC_ICAdd_GotoIf(U8*,U8*)
 extern U8 *__HC_ICAdd_Str(U8*,U8*); //Takes a AIWNIOS code misc
@@ -1354,6 +1361,8 @@ U8 *AiwniosCompile(CCmpCtrl *cc,I64 *res_sz=NULL,CDbgInfo **info,Bool byte_code=
   Bool force_segment=FALSE;
   I64 statics_sz=0;
   static I64 static_cnt=0;
+  if (cc->htc.fun && cc->htc.fun->flags&FSF_INTERRUPT)
+    __HC_CmpCtrl_SetInterrupt(acc,cc->htc.fun->flags&FSF_HASERRCODE);
   stat_nam=MStrPrint(".STATIC_%d",static_cnt++);
 //Turn all IC_IMM_I64s that are used as F64's to IC_IMM_I64
   for(cur=head->next;cur!=head;cur=cur->next) {
@@ -1421,6 +1430,12 @@ skip_convert:;
   #endif
   if(cc->flags&CCF_AOT_COMPILE)
     __HC_CmpCtrl_SetAOT(acc);
+  if (cc->flags&CCF_AOT_COMPILE &&
+      (Bt(&cc->opts,OPTf_RAW_AOT) ||
+       Bt(&cc->opts,OPTf_TEMPLEOS_BIN_SIGNATURE)))
+    __HC_CmpCtrl_UseNativeSegments(acc);
+  if (cc->htc.fun && Bt(&cc->htc.fun->flags,Ff_DOT_DOT_DOT))
+    __HC_CmpCtrl_SetVargs(acc);
   if(cc->flags&CCF_RAW_AOT_ASM)
     __HC_CmpCtrl_SetRawAOTAsm(acc);
   if(info) {
@@ -1509,7 +1524,9 @@ skip_convert:;
         if(foff&(align-1))
           foff+=align-(foff&(align-1)); //Align to 8
         #else 
-	if(IsFunArg(cc->htc.fun,mlst)&&!(mlst->flags&MLF_DOT_DOT_DOT&&!StrCmp(mlst->str,"argv"))) {
+	if(IsFunArg(cc->htc.fun,mlst) &&
+              (!(mlst->flags&MLF_DOT_DOT_DOT&&!StrCmp(mlst->str,"argv")) ||
+               Bt(&cc->opts,OPTf_TEMPLEOS_BIN_SIGNATURE))) {
 	  //Arguments are passed on stack so no need to mode
 	  foff_lookup[idx].i32[0]=mlst->offset;
 	  //16 to offset [PC,RBP] pair at base 
@@ -1588,7 +1605,8 @@ found_offset:;;
 #ifdef TARGET_X86
 //Not yet implemented for AARCH64
       if(mlst->flags&MLF_DOT_DOT_DOT&&!StrCmp(mlst->str,"argv")) {
-        __HC_ICAdd_GetVargsPtr(cc2);
+        if (!Bt(&cc->opts,OPTf_TEMPLEOS_BIN_SIGNATURE))
+          __HC_ICAdd_GetVargsPtr(cc2);
       } else
         __HC_ICAdd_Arg(cc2,idx);
 #endif
@@ -1725,8 +1743,13 @@ found_offset:;;
     #endif
     if(force_segment) {
       //Bypass import and use raw one on X86,they are predictable
-      misc->h=HashFind("__Fs",Fs->hash_table,HTT_FUN);
-      new=__HC_ICAdd_I64(cc2,misc->h(CHashFun*)->exe_addr);      
+      if(Bt(&cc->opts,OPTf_RAW_AOT) ||
+            Bt(&cc->opts,OPTf_TEMPLEOS_BIN_SIGNATURE))
+        new=__HC_ICAdd_I64(cc2,0);
+      else {
+        misc->h=HashFind("__Fs",Fs->hash_table,HTT_FUN);
+        new=__HC_ICAdd_I64(cc2,misc->h(CHashFun*)->exe_addr);
+      }
     } else if(!(misc->h->type&HTF_IMPORT)) {
       new=__HC_ICAdd_I64(cc2,misc->h(CHashFun*)->exe_addr);
     } else
@@ -1759,8 +1782,13 @@ found_offset:;;
     #endif
     if(force_segment) {
       //Bypass import and use raw one on X86,they are predictable
-      misc->h=HashFind("__Gs",Fs->hash_table,HTT_FUN);
-      new=__HC_ICAdd_I64(cc2,misc->h(CHashFun*)->exe_addr);      
+      if(Bt(&cc->opts,OPTf_RAW_AOT) ||
+            Bt(&cc->opts,OPTf_TEMPLEOS_BIN_SIGNATURE))
+        new=__HC_ICAdd_I64(cc2,0);
+      else {
+        misc->h=HashFind("__Gs",Fs->hash_table,HTT_FUN);
+        new=__HC_ICAdd_I64(cc2,misc->h(CHashFun*)->exe_addr);
+      }
     } else if(!(misc->h->type&HTF_IMPORT)) {
       new=__HC_ICAdd_I64(cc2,misc->h(CHashFun*)->exe_addr);
     } else
@@ -1831,6 +1859,11 @@ found_offset:;;
     break;case IC_NOP2:
     break;case IC_LABEL:
       new=__HC_ICAdd_Label(cc2,cur->ic_data(CCodeMisc*)->addr);
+    break;case IC_GET_LABEL:
+      misc=cur->ic_data;
+      if (!misc->addr)
+        misc->addr=__HC_CodeMiscLabelNew(acc,&misc->addr2);
+      new=__HC_ICAdd_LabelAddr(cc2,misc->addr);
 break;case IC_ASM:
   __HC_ICAdd_Label(cc2,__HC_CodeMiscLabelNew(acc,&cur->ic_data(CAiwniosAOTBlob*)->final_start_addr));
    new=__HC_ICAdd_RawBytes(cc2,cur->ic_data(CAiwniosAOTBlob*)->data,cur->ic_data(CAiwniosAOTBlob*)->len);
@@ -2077,7 +2110,10 @@ break;case IC_RIP:
     break;case IC_SUB_EQU:
     new=__HC_ICAdd_SubEq(cc2);
     break;case IC_JMP:
-    new=__HC_ICAdd_Goto(cc2,cur->ic_data(CCodeMisc*)->addr);
+    misc=cur->ic_data;
+    if (!misc->addr)
+      misc->addr=__HC_CodeMiscLabelNew(acc,&misc->addr2);
+    new=__HC_ICAdd_Goto(cc2,misc->addr);
     break;case IC_SUB_RET:
     new=__HC_ICAdd_SubRet(cc2);
     break;case IC_SUB_CALL:
diff --git a/Src/AIWNIOS_PrsExp.HC b/Src/AIWNIOS_PrsExp.HC
index 1653670..823cbc1 100644
--- a/Src/AIWNIOS_PrsExp.HC
+++ b/Src/AIWNIOS_PrsExp.HC
@@ -86,8 +86,7 @@ CHashClass *AssignRawTypeToNode(CRPN *r) {
     return r->ic_class=cmp.internal_types[RT_I64];
     break;case IC_TYPE:
     goto unimp;
-    break;case IC_GET_LABEL:
-    goto unimp;
+    break;case IC_GET_LABEL: return r->ic_class;
     break;case IC_REG:
     goto unimp;
     break;case IC_TO_I64: return r->ic_class=cmp.internal_types[RT_I64];
diff --git a/Src/CInit.HC b/Src/CInit.HC
index f2f36cd..3c0806d 100644
--- a/Src/CInit.HC
+++ b/Src/CInit.HC
@@ -426,7 +426,7 @@ U0 CMiscInit() {
   CmpAddKw(cmp.asm_hash,"static",41);
   CmpAddKw(cmp.asm_hash,"lock",42);
   CmpAddKw(cmp.asm_hash,"defined",43);
-  CmpAddKw(cmp.asm_hash,"interupt",44);
+  CmpAddKw(cmp.asm_hash,"interrupt",44);
   CmpAddKw(cmp.asm_hash,"haserrcode",45);
   CmpAddKw(cmp.asm_hash,"argpop",46);
   CmpAddKw(cmp.asm_hash,"noargpop",47);
diff --git a/Src/PrsStmt.HC b/Src/PrsStmt.HC
index 0a6f2f6..e312cde 100644
--- a/Src/PrsStmt.HC
+++ b/Src/PrsStmt.HC
@@ -103,10 +103,10 @@ CHashFun *PrsFunJoin(CCmpCtrl *cc,CHashClass *tmp_return,
     else
       tmpf->exe_addr=&UndefinedExtern;
     LBts(&tmpf->flags,Cf_EXTERN);
-    tmpf->flags|=fsp_flags&FSG_FUN_FLAGS1;
     if (name) //if not fun_ptr
       HashAdd(tmpf,cc->htc.glbl_hash_table);
   }
+  tmpf->flags|=fsp_flags&FSG_FUN_FLAGS1;
   if(fsp_flags&FSF_PUBLIC)
     LBts(&tmpf->type,HTf_PUBLIC);
   else
@@ -953,16 +953,13 @@ U0 PrsTryBlk(CCmpCtrl *cc,I64 try_cnt)
   CHashClass	*tmpc=cmp.internal_types[RT_PTR];
   CHashFun	*tmp_try=HashFind("SysTry",cc->htc.hash_table_lst,HTT_FUN),
 	*tmp_untry=HashFind("SysUntry",cc->htc.hash_table_lst,HTT_FUN),
-  *tmp_a_ent=HashFind("AIWNIOS_SetJmp",cc->htc.hash_table_lst,HTT_FUN),
-  *tmp_end_catch=HashFind("EndCatch",cc->htc.hash_table_lst,HTT_FUN);
+  *tmp_a_ent,*tmp_end_catch;
 
-  if (!tmp_try || !tmp_untry || !tmp_a_ent)
-    LexExcept(cc,"Missing header for SysTry(),EndCatch(),AIWNIOS_SetJmp() and SysUntry() at ");
+  if (!tmp_try || !tmp_untry)
+    LexExcept(cc,"Missing header for SysTry() and SysUntry() at ");
 
   tmp_try=CodeMiscHashNew(cc,tmp_try);
   tmp_untry=CodeMiscHashNew(cc,tmp_untry);
-  tmp_a_ent=CodeMiscHashNew(cc,tmp_a_ent);
-  tmp_end_catch=CodeMiscHashNew(cc,tmp_end_catch);
   
   cc->flags|=CCF_NO_REG_OPT; //TODO:Currently no reg vars in funs with try/catch
 
@@ -970,6 +967,41 @@ U0 PrsTryBlk(CCmpCtrl *cc,I64 try_cnt)
   lb_done =COCMiscNew(cc,CMT_LABEL);
   lb_untry=COCMiscNew(cc,CMT_LABEL);
 
+  if (cc->flags&CCF_AOT_COMPILE &&
+      (Bt(&cc->opts,OPTf_RAW_AOT) ||
+       Bt(&cc->opts,OPTf_TEMPLEOS_BIN_SIGNATURE))) {
+    RPNAddNew(cc,IC_ADDR_IMPORT,tmp_try,tmpc,FALSE);
+    RPNAddNew(cc,IC_GET_LABEL,lb_catch,tmpc,FALSE);
+    RPNAddNew(cc,IC_GET_LABEL,lb_untry,tmpc,FALSE);
+    RPNAddNew(cc,IC_CALL_INDIRECT,3,tmpc,FALSE);
+
+    PrsStmt(cc,try_cnt+1);
+
+    RPNAddNew(cc,IC_LABEL,lb_untry,0,FALSE);
+    RPNAddNew(cc,IC_ADDR_IMPORT,tmp_untry,tmpc,FALSE);
+    RPNAddNew(cc,IC_CALL_INDIRECT,1,tmpc,FALSE);
+    RPNAddNew(cc,IC_JMP,lb_done,0,FALSE);
+
+    if (PrsKeyWord(cc)!=KW_CATCH)
+      LexExcept(cc,"Missing 'catch' at");
+
+    Lex(cc);
+    RPNAddNew(cc,IC_LABEL,lb_catch,0,FALSE);
+    PrsStmt(cc,try_cnt+1);
+    RPNAddNew(cc,IC_IMM_I64,0,cmp.internal_types[RT_I64],FALSE);
+    RPNAddNew(cc,IC_RET,0,tmpc,FALSE);
+    RPNAddNew(cc,IC_LABEL,lb_done,0,FALSE);
+    return;
+  }
+
+  tmp_a_ent=HashFind("AIWNIOS_SetJmp",cc->htc.hash_table_lst,HTT_FUN);
+  tmp_end_catch=HashFind("EndCatch",cc->htc.hash_table_lst,HTT_FUN);
+  if (!tmp_a_ent || !tmp_end_catch)
+    LexExcept(cc,"Missing header for AIWNIOS_SetJmp() and EndCatch() at ");
+
+  tmp_a_ent=CodeMiscHashNew(cc,tmp_a_ent);
+  tmp_end_catch=CodeMiscHashNew(cc,tmp_end_catch);
+
   //Changed by nroot
   RPNAddNew(cc,IC_ADDR_IMPORT,tmp_a_ent,tmpc,FALSE);
   RPNAddNew(cc,IC_ADDR_IMPORT,tmp_try,tmpc,FALSE);
diff --git a/Src/PrsVar.HC b/Src/PrsVar.HC
index 6af5374..ee1ddb8 100644
--- a/Src/PrsVar.HC
+++ b/Src/PrsVar.HC
@@ -404,13 +404,22 @@ U0 PrsDotDotDot(CCmpCtrl *cc,CHashFun *tmpf,I64 _reg)
 
   tmpm=MemberLstNew(_reg);
   tmpm->flags=MLF_DOT_DOT_DOT;
-  tmpm->member_class=cmp.internal_types[RT_I64]; //Changed by nroot,argv is a pointer,not an array
-  tmpm->member_class++;
+  if (Bt(&cc->opts,OPTf_TEMPLEOS_BIN_SIGNATURE)) {
+    tmpm->member_class=cmp.internal_types[RT_I64];
+    tmpm->dim.total_cnt=127; //arbitrary
+    tmpm->dim.next=MAlloc(sizeof(CArrayDim));
+    tmpm->dim.next->next=NULL;
+    tmpm->dim.next->cnt=127;
+    tmpm->dim.next->total_cnt=1;
+  } else {
+    tmpm->member_class=cmp.internal_types[RT_I64];
+    tmpm->member_class++;
+    tmpm->dim.total_cnt=1;
+  }
   tmpm->str=StrNew("argv");
   tmpm->offset=tmpf->size;
-  tmpm->size=1; //Close enough(1 count of an I8* ???)
-  tmpf->size+=8;//Close enough
-  tmpm->dim.total_cnt=1; //Added by nroot
+  tmpm->size=8;
+  tmpf->size+=8;
   MemberAdd(cc,tmpm,tmpf,PRS1B_FUN_ARG);
 
   if (cc->token==')')
diff --git a/c/aiwn_lexparser.h b/c/aiwn_lexparser.h
index 0854a0f..56ea885 100644
--- a/c/aiwn_lexparser.h
+++ b/c/aiwn_lexparser.h
@@ -186,6 +186,10 @@ typedef struct CCmpCtrl {
 #define CCF_AOT_COMPILE              0x4
 #define CCF_ICMOV_NO_USE_RAX         0x8
 #define CCF_RAW_AOT_ASM              0x10
+#define CCF_NATIVE_SEGMENTS          0x20
+#define CCF_INTERRUPT                0x40
+#define CCF_INTERRUPT_ERROR_CODE     0x80
+#define CCF_VARGS                    0x100
   int64_t flags;
   CHashFun *cur_fun;
   CCodeCtrl *code_ctrl;
@@ -728,6 +732,7 @@ char *__HC_Compile(CCmpCtrl *ccmp, int64_t *sz, char **dbg_info,
                    struct CHeapCtrl *heap);
 CRPN *__HC_ICAdd_Goto(CCodeCtrl *cc, CCodeMisc *cm);
 CRPN *__HC_ICAdd_GotoIf(CCodeCtrl *cc, CCodeMisc *cm);
+CRPN *__HC_ICAdd_LabelAddr(CCodeCtrl *cc, CCodeMisc *misc);
 CRPN *__HC_ICAdd_Str(CCodeCtrl *cc, CCodeMisc *cm);
 CRPN *__HC_ICAdd_Label(CCodeCtrl *cc, CCodeMisc *misc);
 CRPN *__HC_ICAdd_Arg(CCodeCtrl *cc, int64_t arg);
@@ -766,6 +771,9 @@ extern void PrsBindCSymbolNaked(char *name, void *ptr, int64_t arity);
 extern void PrsBindCSymbol(char *name, void *ptr, int64_t arity);
 void __HC_CmpCtrl_SetAOT(CCmpCtrl *cc);
 void __HC_CmpCtrl_SetRawAOTAsm(CCmpCtrl *cc);
+void __HC_CmpCtrl_UseNativeSegments(CCmpCtrl *cc);
+void __HC_CmpCtrl_SetVargs(CCmpCtrl *cc);
+void __HC_CmpCtrl_SetInterrupt(CCmpCtrl *cc, int64_t has_error_code);
 
 CRPN *__HC_ICAdd_GetVargsPtr(CCodeCtrl *cc);
 
diff --git a/c/arm_backend.c b/c/arm_backend.c
index 11ee9fc..400d1c2 100644
--- a/c/arm_backend.c
+++ b/c/arm_backend.c
@@ -2537,6 +2537,10 @@ static int64_t FuncProlog(CCmpCtrl *cctrl, char *bin, int64_t code_off) {
   CMemberLst *lst;
   CICArg fun_arg = {0}, write_to = {0};
   CRPN *rpn, *arg;
+  if (cctrl->flags & CCF_INTERRUPT) {
+    fprintf(stderr, "interrupt functions are unsupported on AArch64\n");
+    abort();
+  }
   /* Old SP
    * saved regs
    * wiggle room
diff --git a/c/main.c b/c/main.c
index a8d39cb..1294dc8 100644
--- a/c/main.c
+++ b/c/main.c
@@ -923,6 +923,21 @@ static int64_t STK___HC_CmpCtrl_SetRawAOTAsm(int64_t *stk) {
   return 0;
 }
 
+static int64_t STK___HC_CmpCtrl_UseNativeSegments(int64_t *stk) {
+  __HC_CmpCtrl_UseNativeSegments((CCmpCtrl *)stk[0]);
+  return 0;
+}
+
+static int64_t STK___HC_CmpCtrl_SetVargs(int64_t *stk) {
+  __HC_CmpCtrl_SetVargs((CCmpCtrl *)stk[0]);
+  return 0;
+}
+
+static int64_t STK___HC_CmpCtrl_SetInterrupt(int64_t *stk) {
+  __HC_CmpCtrl_SetInterrupt((CCmpCtrl *)stk[0], stk[1]);
+  return 0;
+}
+
 static int64_t STK__HC_ICAdd_RawBytes(int64_t *stk) {
   return (int64_t)__HC_ICAdd_RawBytes((CCmpCtrl *)stk[0], (char *)stk[1],
                                       stk[2]);
@@ -1156,6 +1171,9 @@ static int64_t STK___HC_CodeMiscStrNew(int64_t *stk) {
 static int64_t STK___HC_CodeMiscLabelNew(int64_t *stk) {
   return (int64_t)__HC_CodeMiscLabelNew((CCmpCtrl *)stk[0], (void **)stk[1]);
 }
+static int64_t STK___HC_ICAdd_LabelAddr(int64_t *stk) {
+  return (int64_t)__HC_ICAdd_LabelAddr((CCodeCtrl *)stk[0], (CCodeMisc *)stk[1]);
+}
 static int64_t STK___HC_CmpCtrlNew(int64_t *stk) {
   return (int64_t)__HC_CmpCtrlNew();
 }
@@ -1735,6 +1753,11 @@ static void BootAiwnios(char *bootstrap_text) {
     PrsAddSymbol("IsValidPtr", STK_IsValidPtr, 1);
     PrsAddSymbol("__HC_CmpCtrl_SetAOT", STK___HC_CmpCtrl_SetAOT, 1);
     PrsAddSymbol("__HC_CmpCtrl_SetRawAOTAsm", STK___HC_CmpCtrl_SetRawAOTAsm, 1);
+    PrsAddSymbol("__HC_CmpCtrl_UseNativeSegments",
+                 STK___HC_CmpCtrl_UseNativeSegments, 1);
+    PrsAddSymbol("__HC_CmpCtrl_SetVargs", STK___HC_CmpCtrl_SetVargs, 1);
+    PrsAddSymbol("__HC_CmpCtrl_SetInterrupt", STK___HC_CmpCtrl_SetInterrupt,
+                 2);
     PrsAddSymbol("__HC_ICAdd_Typecast", STK___HC_ICAdd_Typecast, 3);
     PrsAddSymbol("__HC_ICAdd_SubCall", STK___HC_ICAdd_SubCall, 2);
     PrsAddSymbol("__HC_ICAdd_SubProlog", STK___HC_ICAdd_SubProlog, 1);
@@ -1796,6 +1819,7 @@ static void BootAiwnios(char *bootstrap_text) {
     PrsAddSymbol("__HC_CodeMiscStrNew", STK___HC_CodeMiscStrNew, 3);
     PrsAddSymbol("__HC_CodeMiscJmpTableNew", STK___HC_CodeMiscJmpTableNew, 4);
     PrsAddSymbol("__HC_ICAdd_Label", STK___HC_ICAdd_Label, 2);
+    PrsAddSymbol("__HC_ICAdd_LabelAddr", STK___HC_ICAdd_LabelAddr, 2);
     PrsAddSymbol("__HC_ICAdd_Goto", STK___HC_ICAdd_Goto, 2);
     PrsAddSymbol("__HC_ICAdd_GotoIf", STK___HC_ICAdd_GotoIf, 2);
     PrsAddSymbol("__HC_ICAdd_Str", STK___HC_ICAdd_Str, 2);
diff --git a/c/optpass.c b/c/optpass.c
index 47057dd..65637f7 100644
--- a/c/optpass.c
+++ b/c/optpass.c
@@ -33,9 +33,12 @@ static void FixFunArgs(CCmpCtrl *cctrl, CRPN *rpn) {
   CHashFun *ic_fun;
   CMemberLst *arg, *args_lst_rev;
   for (cnt = 0; cnt != rpn->length; cnt++)
-    rpn2 = ICFwd(rpn2);
+    if (!(rpn2 = ICFwd(rpn2)))
+      return;
   AssignRawTypeToNode(cctrl, rpn2);
   ic_fun = rpn2->ic_fun;
+  if (!ic_fun)
+    return;
   rpn2 = rpn->base.next;
   arg = ic_fun->base.members_lst;
   cnt2 = ic_fun->argc;
diff --git a/c/parser.c b/c/parser.c
index 1454873..c7ec9fa 100644
--- a/c/parser.c
+++ b/c/parser.c
@@ -4923,6 +4923,18 @@ CRPN *__HC_ICAdd_Label(CCodeCtrl *cc, CCodeMisc *misc) {
   return rpn;
 }
 
+CRPN *__HC_ICAdd_LabelAddr(CCodeCtrl *cc, CCodeMisc *misc) {
+  CRPN *rpn;
+  *(rpn = A_CALLOC(sizeof(CRPN), cc->hc)) = (CRPN){
+      .type = IC_SHORT_ADDR,
+      .code_misc = misc,
+  };
+  rpn->ic_class = HashFind("U8i", Fs->hash_table, HTT_CLASS, 1) + 1;
+  rpn->raw_type = RT_PTR;
+  QueIns(rpn, cc->ir_code);
+  return rpn;
+}
+
 CRPN *__HC_ICAdd_Goto(CCodeCtrl *cc, CCodeMisc *cm) {
   CRPN *rpn;
   *(rpn = A_CALLOC(sizeof(CRPN), cc->hc)) = (CRPN){
@@ -5086,6 +5098,20 @@ void __HC_CmpCtrl_SetAOT(CCmpCtrl *cc) {
 void __HC_CmpCtrl_SetRawAOTAsm(CCmpCtrl *cc) {
   cc->flags |= CCF_RAW_AOT_ASM;
 }
+
+void __HC_CmpCtrl_UseNativeSegments(CCmpCtrl *cc) {
+  cc->flags |= CCF_NATIVE_SEGMENTS;
+}
+
+void __HC_CmpCtrl_SetVargs(CCmpCtrl *cc) {
+  cc->flags |= CCF_VARGS;
+}
+
+void __HC_CmpCtrl_SetInterrupt(CCmpCtrl *cc, int64_t has_error_code) {
+  cc->flags |= CCF_INTERRUPT;
+  if (has_error_code)
+    cc->flags |= CCF_INTERRUPT_ERROR_CODE;
+}
 HC_IC_BINDING(HC_ICAdd_ToBool, IC_TO_BOOL);
 HC_IC_BINDING(HC_ICAdd_BT, IC_BT);
 HC_IC_BINDING(HC_ICAdd_BTC, IC_BTC);
diff --git a/c/riscv_backend.c b/c/riscv_backend.c
index fed1468..a9c2c01 100644
--- a/c/riscv_backend.c
+++ b/c/riscv_backend.c
@@ -2364,6 +2364,10 @@ static int64_t FuncProlog(CCmpCtrl *cctrl, char *bin, int64_t code_off) {
   CMemberLst *lst;
   CICArg fun_arg = {0}, write_to = {0}, stack = {0}, tmp = {0};
   CRPN *rpn, *arg;
+  if (cctrl->flags & CCF_INTERRUPT) {
+    fprintf(stderr, "interrupt functions are unsupported on RISC-V\n");
+    abort();
+  }
   if (cctrl->cur_fun) {
     fsz = cctrl->cur_fun->base.sz + 16; //+16 for RBP/return address
     arg_cnt = cctrl->cur_fun->argc;
diff --git a/c/x86_64_backend.c b/c/x86_64_backend.c
index 86f9c12..0907eb1 100644
--- a/c/x86_64_backend.c
+++ b/c/x86_64_backend.c
@@ -387,6 +387,16 @@ int64_t X86Ret(char *to, int64_t ul) {
   return len;
 }
 
+static int64_t X86Iretq(char *to, int64_t unused) {
+  int64_t len = 0;
+  char buf[16];
+  if (!to)
+    to = buf;
+  ADD_U8(0x48);
+  ADD_U8(0xcf);
+  return len;
+}
+
 int64_t X86MovImm(char *to, int64_t a, int64_t off) {
   int64_t len = 0;
   char buf[16];
@@ -5293,6 +5303,15 @@ static int64_t __FindPushedIRegs(CCmpCtrl *cctrl, char *to_push) {
   int64_t cnt = 0;
   int64_t idx = 0;
   memset(to_push, 0, 16);
+  if (cctrl->flags & CCF_INTERRUPT) {
+    for (idx = 0; idx != 16; idx++) {
+      if (idx != RSP && idx != RBP) {
+        to_push[idx] = 1;
+        cnt++;
+      }
+    }
+    return cnt;
+  }
   if (!cctrl->cur_fun) {
     if (!cctrl->found_used_iregs) {
       // Perhaps we are being called from HolyC and we aren't using a "cur_fun"
@@ -5697,6 +5716,7 @@ static int64_t FuncEpilog(CCmpCtrl *cctrl, char *bin, int64_t code_off) {
     arg_cnt = cctrl->cur_fun->argc;
     is_vargs = !!(cctrl->cur_fun->base.flags & CLSF_VARGS);
   } else {
+    is_vargs = !!(cctrl->flags & CCF_VARGS);
     fsz = 16;
     for (rpn = cctrl->code_ctrl->ir_code->next;
          rpn != cctrl->code_ctrl->ir_code; rpn = rpn->base.next) {
@@ -5743,7 +5763,11 @@ static int64_t FuncEpilog(CCmpCtrl *cctrl, char *bin, int64_t code_off) {
 
   off = -old_regs_start;
   AIWNIOS_ADD_CODE(X86Leave, 0);
-  if (is_vargs) {
+  if (cctrl->flags & CCF_INTERRUPT) {
+    if (cctrl->flags & CCF_INTERRUPT_ERROR_CODE)
+      AIWNIOS_ADD_CODE(X86AddImm32, RSP, 8);
+    AIWNIOS_ADD_CODE(X86Iretq, 0);
+  } else if (is_vargs) {
     AIWNIOS_ADD_CODE(X86Ret, 0);
   } else {
     AIWNIOS_ADD_CODE(X86Ret, 8 * arg_cnt);
@@ -6387,13 +6411,20 @@ int64_t __OptPassFinal(CCmpCtrl *cctrl, CRPN *rpn, char *bin,
       abort();
     }
   segment:
-    // %gs is used as a TLS reg for all OS'es
     into_reg = 0;
+    if (cctrl->flags & CCF_NATIVE_SEGMENTS) {
+      if (rpn->type == IC_FS) {
+        AIWNIOS_ADD_CODE(SEG_FS, 0);
+      } else {
+        AIWNIOS_ADD_CODE(SEG_GS, 0);
+      }
+    } else {
 #if __OpenBSD__ + __NetBSD__ > 0
-    AIWNIOS_ADD_CODE(SEG_FS, 0);
+      AIWNIOS_ADD_CODE(SEG_FS, 0);
 #else
-    AIWNIOS_ADD_CODE(SEG_GS, 0);
+      AIWNIOS_ADD_CODE(SEG_GS, 0);
 #endif
+    }
     if (rpn->res.mode == MD_REG && next->type == IC_I64 &&
         Is32Bit(next->integer)) {
       into_reg = rpn->res.reg;
-- 
2.53.0
